File: packages/musical-conductor/modules/communication/sequences/monitoring/PerformanceTracker.ts
/**
 * PerformanceTracker - Performance timing and beat execution tracking
 * Handles beat timing, execution duration tracking, and performance monitoring
 */

export interface BeatTiming {
  sequenceName: string;
  beat: number;
  startTime: number;
  endTime?: number;
  duration?: number;
}

export interface SequenceTiming {
  sequenceName: string;
  startTime: number;
  endTime?: number;
  duration?: number;
  beatCount: number;
}

export interface MovementTiming {
  sequenceName: string;
  movementName: string;
  requestId: string;
  startTime: number;
  endTime?: number;
  duration?: number;
  beatsCount: number;
}

export class PerformanceTracker {
  private beatStartTimes: Map<string, number> = new Map();
  private sequenceTimings: Map<string, SequenceTiming> = new Map();
  private movementStartTimes: Map<string, number> = new Map();
  private completedBeats: BeatTiming[] = [];
  private completedMovements: MovementTiming[] = [];
  private maxHistorySize: number = 1000;

  /**
   * Start tracking a beat execution
   * @param sequenceName - Sequence name
   * @param beat - Beat number
   * @returns Beat key for tracking
   */
  startBeatTiming(sequenceName: string, beat: number): string {
    const beatKey = `${sequenceName}-${beat}`;
    const startTime = performance.now();
    this.beatStartTimes.set(beatKey, startTime);

    console.log(
      `⏱️ PerformanceTracker: Started timing beat ${beat} for ${sequenceName}`
    );

    return beatKey;
  }

  /**
   * End tracking a beat execution
   * @param sequenceName - Sequence name
   * @param beat - Beat number
   * @returns Beat duration in milliseconds
   */
  endBeatTiming(sequenceName: string, beat: number): number | null {
    const beatKey = `${sequenceName}-${beat}`;
    const startTime = this.beatStartTimes.get(beatKey);

    if (!startTime) {
      console.warn(
        `⏱️ PerformanceTracker: No start time found for beat ${beat} in ${sequenceName}`
      );
      return null;
    }

    const endTime = performance.now();
    const duration = endTime - startTime;

    // Record the completed beat
    const beatTiming: BeatTiming = {
      sequenceName,
      beat,
      startTime,
      endTime,
      duration,
    };

    this.completedBeats.push(beatTiming);
    this.beatStartTimes.delete(beatKey);

    // Maintain history size limit
    if (this.completedBeats.length > this.maxHistorySize) {
      this.completedBeats = this.completedBeats.slice(-this.maxHistorySize / 2);
    }

    console.log(
      `⏱️ PerformanceTracker: Beat ${beat} completed in ${duration.toFixed(
        2
      )}ms`
    );

    return duration;
  }

  /**
   * Clean up timing data for a failed beat
   * @param sequenceName - Sequence name
   * @param beat - Beat number
   */
  cleanupFailedBeat(sequenceName: string, beat: number): void {
    const beatKey = `${sequenceName}-${beat}`;
    this.beatStartTimes.delete(beatKey);

    console.warn(
      `⏱️ PerformanceTracker: Cleaned up failed beat ${beat} for ${sequenceName}`
    );
  }

  /**
   * Start tracking a movement execution
   * @param sequenceName - Sequence name
   * @param movementName - Movement name
   * @param requestId - Request identifier
   * @returns Movement key for tracking
   */
  startMovementTiming(
    sequenceName: string,
    movementName: string,
    requestId: string
  ): string {
    const movementKey = `${sequenceName}-${movementName}-${requestId}`;
    const startTime = performance.now();
    this.movementStartTimes.set(movementKey, startTime);

    console.log(
      `⏱️ PerformanceTracker: Started timing movement ${movementName} for ${sequenceName}`
    );

    return movementKey;
  }

  /**
   * End tracking a movement execution
   * @param sequenceName - Sequence name
   * @param movementName - Movement name
   * @param requestId - Request identifier
   * @param beatsCount - Number of beats executed in the movement
   * @returns Movement duration in milliseconds
   */
  endMovementTiming(
    sequenceName: string,
    movementName: string,
    requestId: string,
    beatsCount: number
  ): number | null {
    const movementKey = `${sequenceName}-${movementName}-${requestId}`;
    const startTime = this.movementStartTimes.get(movementKey);

    if (!startTime) {
      console.warn(
        `⏱️ PerformanceTracker: No start time found for movement ${movementName} in ${sequenceName}`
      );
      return null;
    }

    const endTime = performance.now();
    const duration = endTime - startTime;

    // Record the completed movement
    const movementTiming: MovementTiming = {
      sequenceName,
      movementName,
      requestId,
      startTime,
      endTime,
      duration,
      beatsCount,
    };

    this.completedMovements.push(movementTiming);
    this.movementStartTimes.delete(movementKey);

    // Maintain history size limit
    if (this.completedMovements.length > this.maxHistorySize) {
      this.completedMovements = this.completedMovements.slice(
        -this.maxHistorySize / 2
      );
    }

    console.log(
      `⏱️ PerformanceTracker: Movement ${movementName} completed in ${duration.toFixed(
        2
      )}ms`
    );

    return duration;
  }

  /**
   * Clean up timing data for a failed movement
   * @param sequenceName - Sequence name
   * @param movementName - Movement name
   * @param requestId - Request identifier
   */
  cleanupFailedMovement(
    sequenceName: string,
    movementName: string,
    requestId: string
  ): void {
    const movementKey = `${sequenceName}-${movementName}-${requestId}`;
    this.movementStartTimes.delete(movementKey);

    console.warn(
      `⏱️ PerformanceTracker: Cleaned up failed movement ${movementName} for ${sequenceName}`
    );
  }

  /**
   * Start tracking a sequence execution
   * @param sequenceName - Sequence name
   * @param executionId - Execution identifier
   */
  startSequenceTiming(sequenceName: string, executionId: string): void {
    const startTime = performance.now();
    this.sequenceTimings.set(executionId, {
      sequenceName,
      startTime,
      beatCount: 0,
    });

    console.log(
      `⏱️ PerformanceTracker: Started timing sequence ${sequenceName} (${executionId})`
    );
  }

  /**
   * End tracking a sequence execution
   * @param executionId - Execution identifier
   * @returns Sequence duration in milliseconds
   */
  endSequenceTiming(executionId: string): number | null {
    const timing = this.sequenceTimings.get(executionId);

    if (!timing) {
      console.warn(
        `⏱️ PerformanceTracker: No timing found for execution ${executionId}`
      );
      return null;
    }

    const endTime = performance.now();
    const duration = endTime - timing.startTime;

    timing.endTime = endTime;
    timing.duration = duration;

    console.log(
      `⏱️ PerformanceTracker: Sequence ${
        timing.sequenceName
      } completed in ${duration.toFixed(2)}ms`
    );

    // Clean up after recording
    this.sequenceTimings.delete(executionId);

    return duration;
  }

  /**
   * Increment beat count for a sequence
   * @param executionId - Execution identifier
   */
  incrementBeatCount(executionId: string): void {
    const timing = this.sequenceTimings.get(executionId);
    if (timing) {
      timing.beatCount++;
    }
  }

  /**
   * Get current beat timing information
   * @param sequenceName - Sequence name
   * @param beat - Beat number
   * @returns Current timing info or null
   */
  getCurrentBeatTiming(
    sequenceName: string,
    beat: number
  ): { startTime: number; elapsed: number } | null {
    const beatKey = `${sequenceName}-${beat}`;
    const startTime = this.beatStartTimes.get(beatKey);

    if (!startTime) {
      return null;
    }

    return {
      startTime,
      elapsed: performance.now() - startTime,
    };
  }

  /**
   * Get beat performance statistics
   * @param sequenceName - Optional sequence name filter
   * @returns Beat performance statistics
   */
  getBeatStatistics(sequenceName?: string): {
    totalBeats: number;
    averageDuration: number;
    minDuration: number;
    maxDuration: number;
    recentBeats: BeatTiming[];
  } {
    const filteredBeats = sequenceName
      ? this.completedBeats.filter((b) => b.sequenceName === sequenceName)
      : this.completedBeats;

    if (filteredBeats.length === 0) {
      return {
        totalBeats: 0,
        averageDuration: 0,
        minDuration: 0,
        maxDuration: 0,
        recentBeats: [],
      };
    }

    const durations = filteredBeats
      .map((b) => b.duration!)
      .filter((d) => d !== undefined);
    const averageDuration =
      durations.reduce((sum, d) => sum + d, 0) / durations.length;
    const minDuration = Math.min(...durations);
    const maxDuration = Math.max(...durations);

    return {
      totalBeats: filteredBeats.length,
      averageDuration,
      minDuration,
      maxDuration,
      recentBeats: filteredBeats.slice(-10), // Last 10 beats
    };
  }

  /**
   * Get movement performance statistics
   * @param sequenceName - Optional sequence name filter
   * @returns Movement performance statistics
   */
  getMovementTimings(sequenceName?: string): MovementTiming[] {
    const filteredMovements = sequenceName
      ? this.completedMovements.filter((m) => m.sequenceName === sequenceName)
      : this.completedMovements;

    return filteredMovements.slice(); // Return a copy
  }

  /**
   * Get active timing information
   * @returns Currently active timings
   */
  getActiveTimings(): {
    activeBeats: Array<{ key: string; elapsed: number }>;
    activeSequences: Array<{
      executionId: string;
      sequenceName: string;
      elapsed: number;
    }>;
    activeMovements: Array<{ key: string; elapsed: number }>;
  } {
    const now = performance.now();

    const activeBeats = Array.from(this.beatStartTimes.entries()).map(
      ([key, startTime]) => ({
        key,
        elapsed: now - startTime,
      })
    );

    const activeSequences = Array.from(this.sequenceTimings.entries()).map(
      ([executionId, timing]) => ({
        executionId,
        sequenceName: timing.sequenceName,
        elapsed: now - timing.startTime,
      })
    );

    const activeMovements = Array.from(this.movementStartTimes.entries()).map(
      ([key, startTime]) => ({
        key,
        elapsed: now - startTime,
      })
    );

    return {
      activeBeats,
      activeSequences,
      activeMovements,
    };
  }

  /**
   * Get performance warnings for long-running operations
   * @param beatThreshold - Beat duration threshold in ms (default: 5000)
   * @param sequenceThreshold - Sequence duration threshold in ms (default: 30000)
   * @returns Performance warnings
   */
  getPerformanceWarnings(
    beatThreshold: number = 5000,
    sequenceThreshold: number = 30000
  ): string[] {
    const warnings: string[] = [];
    const now = performance.now();

    // Check for long-running beats
    for (const [beatKey, startTime] of this.beatStartTimes.entries()) {
      const elapsed = now - startTime;
      if (elapsed > beatThreshold) {
        warnings.push(
          `Long-running beat: ${beatKey} (${elapsed.toFixed(0)}ms)`
        );
      }
    }

    // Check for long-running sequences
    for (const [executionId, timing] of this.sequenceTimings.entries()) {
      const elapsed = now - timing.startTime;
      if (elapsed > sequenceThreshold) {
        warnings.push(
          `Long-running sequence: ${timing.sequenceName} (${elapsed.toFixed(
            0
          )}ms)`
        );
      }
    }

    return warnings;
  }

  /**
   * Reset all performance tracking data
   */
  reset(): void {
    this.beatStartTimes.clear();
    this.sequenceTimings.clear();
    this.movementStartTimes.clear();
    this.completedBeats = [];
    this.completedMovements = [];

    console.log("🧹 PerformanceTracker: All tracking data reset");
  }

  /**
   * Get debug information
   * @returns Debug performance information
   */
  getDebugInfo(): {
    activeBeats: number;
    activeSequences: number;
    completedBeatsHistory: number;
    beatStatistics: ReturnType<PerformanceTracker["getBeatStatistics"]>;
    warnings: string[];
  } {
    return {
      activeBeats: this.beatStartTimes.size,
      activeSequences: this.sequenceTimings.size,
      completedBeatsHistory: this.completedBeats.length,
      beatStatistics: this.getBeatStatistics(),
      warnings: this.getPerformanceWarnings(),
    };
  }
}

File: src/MusicalConductor.Avalonia/MusicalConductor.Core/Monitoring/PerformanceTracker.cs
using Microsoft.Extensions.Logging;
using System.Diagnostics;

namespace MusicalConductor.Core.Monitoring;

/// <summary>
/// PerformanceTracker - Performance timing and beat execution tracking
/// Handles beat timing, execution duration tracking, and performance monitoring
/// </summary>
public class PerformanceTracker
{
    public class BeatTiming
    {
        public string SequenceName { get; set; } = string.Empty;
        public int Beat { get; set; }
        public long StartTime { get; set; }
        public long? EndTime { get; set; }
        public double? Duration { get; set; }
    }

    public class SequenceTiming
    {
        public string SequenceName { get; set; } = string.Empty;
        public long StartTime { get; set; }
        public long? EndTime { get; set; }
        public double? Duration { get; set; }
        public int BeatCount { get; set; }
    }

    public class MovementTiming
    {
        public string SequenceName { get; set; } = string.Empty;
        public string MovementName { get; set; } = string.Empty;
        public string RequestId { get; set; } = string.Empty;
        public long StartTime { get; set; }
        public long? EndTime { get; set; }
        public double? Duration { get; set; }
        public int BeatsCount { get; set; }
    }

    private readonly Dictionary<string, long> _beatStartTimes = new();
    private readonly Dictionary<string, SequenceTiming> _sequenceTimings = new();
    private readonly Dictionary<string, long> _movementStartTimes = new();
    private readonly List<BeatTiming> _completedBeats = new();
    private readonly List<MovementTiming> _completedMovements = new();
    private readonly ILogger<PerformanceTracker> _logger;
    private readonly Stopwatch _stopwatch = Stopwatch.StartNew();
    private int _maxHistorySize = 1000;

    public PerformanceTracker(ILogger<PerformanceTracker> logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    /// <summary>
    /// Start tracking a beat execution
    /// </summary>
    public string StartBeatTiming(string sequenceName, int beat)
    {
        var beatKey = $"{sequenceName}-{beat}";
        var startTime = _stopwatch.ElapsedMilliseconds;
        _beatStartTimes[beatKey] = startTime;

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Started timing beat {Beat} for {SequenceName}",
            beat,
            sequenceName);

        if (_logger.IsEnabled(LogLevel.Trace))
        {
            _logger.LogTrace("⏱️ PerformanceTracker: Beat timing details - BeatKey: {BeatKey}, StartTime: {StartTime}ms, TotalBeats: {TotalBeats}", beatKey, startTime, _beatStartTimes.Count);
        }

        return beatKey;
    }

    /// <summary>
    /// End tracking a beat execution
    /// </summary>
    public double? EndBeatTiming(string sequenceName, int beat)
    {
        var beatKey = $"{sequenceName}-{beat}";
        
        if (!_beatStartTimes.TryGetValue(beatKey, out var startTime))
        {
            _logger.LogWarning(
                "⏱️ PerformanceTracker: No start time found for beat {Beat} in {SequenceName}",
                beat,
                sequenceName);
            return null;
        }

        var endTime = _stopwatch.ElapsedMilliseconds;
        var duration = (double)(endTime - startTime);

        // Record the completed beat
        var beatTiming = new BeatTiming
        {
            SequenceName = sequenceName,
            Beat = beat,
            StartTime = startTime,
            EndTime = endTime,
            Duration = duration
        };

        _completedBeats.Add(beatTiming);
        _beatStartTimes.Remove(beatKey);

        // Maintain history size limit
        if (_completedBeats.Count > _maxHistorySize)
        {
            _completedBeats.RemoveRange(0, _maxHistorySize / 2);
        }

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Beat {Beat} completed in {Duration:F2}ms",
            beat,
            duration);

        return duration;
    }

    /// <summary>
    /// Clean up timing data for a failed beat
    /// </summary>
    public void CleanupFailedBeat(string sequenceName, int beat)
    {
        var beatKey = $"{sequenceName}-{beat}";
        _beatStartTimes.Remove(beatKey);

        _logger.LogWarning(
            "⏱️ PerformanceTracker: Cleaned up failed beat {Beat} for {SequenceName}",
            beat,
            sequenceName);

        // Log state transition
        _logger.LogInformation("⚠️ PerformanceTracker: Beat state transitioned to FAILED - {SequenceName}:{Beat}", sequenceName, beat);
    }

    /// <summary>
    /// Start tracking a movement execution
    /// </summary>
    public string StartMovementTiming(string sequenceName, string movementName, string requestId)
    {
        var movementKey = $"{sequenceName}-{movementName}-{requestId}";
        var startTime = _stopwatch.ElapsedMilliseconds;
        _movementStartTimes[movementKey] = startTime;

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Started timing movement {MovementName} for {SequenceName}",
            movementName,
            sequenceName);

        return movementKey;
    }

    /// <summary>
    /// End tracking a movement execution
    /// </summary>
    public double? EndMovementTiming(string sequenceName, string movementName, string requestId, int beatsCount)
    {
        var movementKey = $"{sequenceName}-{movementName}-{requestId}";
        
        if (!_movementStartTimes.TryGetValue(movementKey, out var startTime))
        {
            _logger.LogWarning(
                "⏱️ PerformanceTracker: No start time found for movement {MovementName} in {SequenceName}",
                movementName,
                sequenceName);
            return null;
        }

        var endTime = _stopwatch.ElapsedMilliseconds;
        var duration = (double)(endTime - startTime);

        // Record the completed movement
        var movementTiming = new MovementTiming
        {
            SequenceName = sequenceName,
            MovementName = movementName,
            RequestId = requestId,
            StartTime = startTime,
            EndTime = endTime,
            Duration = duration,
            BeatsCount = beatsCount
        };

        _completedMovements.Add(movementTiming);
        _movementStartTimes.Remove(movementKey);

        // Maintain history size limit
        if (_completedMovements.Count > _maxHistorySize)
        {
            _completedMovements.RemoveRange(0, _maxHistorySize / 2);
        }

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Movement {MovementName} completed in {Duration:F2}ms",
            movementName,
            duration);

        return duration;
    }

    /// <summary>
    /// Clean up timing data for a failed movement
    /// </summary>
    public void CleanupFailedMovement(string sequenceName, string movementName, string requestId)
    {
        var movementKey = $"{sequenceName}-{movementName}-{requestId}";
        _movementStartTimes.Remove(movementKey);

        _logger.LogWarning(
            "⏱️ PerformanceTracker: Cleaned up failed movement {MovementName} for {SequenceName}",
            movementName,
            sequenceName);

        // Log state transition
        _logger.LogInformation("⚠️ PerformanceTracker: Movement state transitioned to FAILED - {SequenceName}:{MovementName}", sequenceName, movementName);
    }

    /// <summary>
    /// Start tracking a sequence execution
    /// </summary>
    public void StartSequenceTiming(string sequenceName)
    {
        var startTime = _stopwatch.ElapsedMilliseconds;
        _sequenceTimings[sequenceName] = new SequenceTiming
        {
            SequenceName = sequenceName,
            StartTime = startTime,
            BeatCount = 0
        };

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Started timing sequence {SequenceName}",
            sequenceName);
    }

    /// <summary>
    /// End tracking a sequence execution
    /// </summary>
    public double? EndSequenceTiming(string sequenceName)
    {
        if (!_sequenceTimings.TryGetValue(sequenceName, out var timing))
        {
            _logger.LogWarning(
                "⏱️ PerformanceTracker: No start time found for sequence {SequenceName}",
                sequenceName);
            return null;
        }

        var endTime = _stopwatch.ElapsedMilliseconds;
        timing.EndTime = endTime;
        timing.Duration = (double)(endTime - timing.StartTime);

        _logger.LogDebug(
            "⏱️ PerformanceTracker: Sequence {SequenceName} completed in {Duration:F2}ms",
            sequenceName,
            timing.Duration.Value);

        return timing.Duration;
    }

    /// <summary>
    /// Get beat statistics for a sequence
    /// </summary>
    public object GetBeatStatistics(string sequenceName)
    {
        var beats = _completedBeats.Where(b => b.SequenceName == sequenceName).ToList();
        
        if (beats.Count == 0)
        {
            return new
            {
                SequenceName = sequenceName,
                TotalBeats = 0,
                AverageDuration = 0.0,
                MinDuration = 0.0,
                MaxDuration = 0.0
            };
        }

        var durations = beats.Select(b => b.Duration ?? 0).ToList();
        
        return new
        {
            SequenceName = sequenceName,
            TotalBeats = beats.Count,
            AverageDuration = durations.Average(),
            MinDuration = durations.Min(),
            MaxDuration = durations.Max()
        };
    }

    /// <summary>
    /// Get movement statistics for a sequence
    /// </summary>
    public object GetMovementStatistics(string sequenceName)
    {
        var movements = _completedMovements.Where(m => m.SequenceName == sequenceName).ToList();
        
        if (movements.Count == 0)
        {
            return new
            {
                SequenceName = sequenceName,
                TotalMovements = 0,
                AverageDuration = 0.0,
                AverageBeatsPerMovement = 0.0
            };
        }

        var durations = movements.Select(m => m.Duration ?? 0).ToList();
        
        return new
        {
            SequenceName = sequenceName,
            TotalMovements = movements.Count,
            AverageDuration = durations.Average(),
            AverageBeatsPerMovement = movements.Average(m => m.BeatsCount)
        };
    }

    /// <summary>
    /// Get all completed beat timings
    /// </summary>
    public IReadOnlyList<BeatTiming> GetCompletedBeats()
    {
        return _completedBeats.AsReadOnly();
    }

    /// <summary>
    /// Get all completed movement timings
    /// </summary>
    public IReadOnlyList<MovementTiming> GetCompletedMovements()
    {
        return _completedMovements.AsReadOnly();
    }

    /// <summary>
    /// Reset all tracking data
    /// </summary>
    public void ResetTrackingData()
    {
        _beatStartTimes.Clear();
        _sequenceTimings.Clear();
        _movementStartTimes.Clear();
        _completedBeats.Clear();
        _completedMovements.Clear();

        _logger.LogInformation("🧹 PerformanceTracker: All tracking data reset");
    }

    /// <summary>
    /// Get debug information
    /// </summary>
    public object GetDebugInfo()
    {
        return new
        {
            ActiveBeatTimings = _beatStartTimes.Count,
            ActiveMovementTimings = _movementStartTimes.Count,
            CompletedBeats = _completedBeats.Count,
            CompletedMovements = _completedMovements.Count,
            MaxHistorySize = _maxHistorySize
        };
    }
}
